### PR TITLE
move ImagePullSecret to ARM

### PR DESCRIPTION
**Pull request by mjudeikis**
```release-note
Move ImagePullSecret to ARM and use combined secret
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -14,6 +15,15 @@ secrets:
 	@mkdir secrets
 	@oc extract -n azure secret/cluster-secrets-azure --to=secrets >/dev/null
 
+private-secrets:
+	@echo "These secrets are sensitive, please do not keep the on your workstation. Execute make private-secrets-clean after you done"
```
Comment (m1kola) [inline Makefile]:
```suggestion
	@echo "These secrets are sensitive, please do not keep them on your workstation. Execute make private-secrets-clean after you done" 
```

File: docs/release.md [hunk 2]
```diff
@@ -71,6 +71,14 @@ Branching model
    with the appropriate version on quay.io. (Note: only members of the
    `openshift-on-azure` quay.io team have access.)
 
+   Images for production are released to ACR. This can be done by people,
+   who has ACR root secrets. These images can be built by checking out
```
Comment (m1kola) [inline docs/release.md]:
```suggestion
   who have ACR root secrets. These images can be built by checking out 
```

**Comment by thekad** [comment]
Wouldn't it be better if you had a target `make acr-push` or similar, that automatically:

1. downloads the private-secrets
2. builds/tags the image
3. pushes the image
4. removes the secrets 

?

Relying on people remembering to run clean target after building is bound to fail

File: docs/alias.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
+# ARO usefull alias
+
+
+## Upload, download privte secrets
```
Comment (asalkeld) [inline docs/alias.md]: s/privte/private

**Comment by jim-minter** [comment]
I am negative on this unless we can absolutely not avoid it.  I'd rather have sane ACR credentials in the first place.

**Comment by mjudeikis** [comment]
| I am negative on this unless we can absolutely not avoid it. I'd rather have sane ACR credentials in the first place.

@jim-minter I think we do have, just our code is not behaving. What do you call sane credential?

**Comment by jim-minter** [comment]
> sane

@mjudeikis something which doesn't allow us to write prod and ideally which does not allow us to overwrite non-RH non-prod either.

**Comment by mjudeikis** [comment]
@jim-minter, we can either wait for this as ACR does not support this,
Or we can ask MSFT to give us different registry and ask them to migrate images (just because we dont trust each other).
Or we can do the same as we did for quay.io (limited access to production secrete) as we did before just with the more stable registry. And this is what this PR delivers.

File: pkg/arm/v10/data/master-startup.sh [hunk 4]
```diff
@@ -28,6 +28,13 @@ yum clean all
 rm -rf /var/lib/yum/client-cert.pem /var/lib/yum/client-key.pem
 {{end}}
 
+# create container pull secret 
+mkdir -p /var/lib/origin/.docker
+cat >/var/lib/origin/.docker/config.json <<'EOF'
+{{ .Derived.CombinedImagePullSecret .Config | String }}
+EOF
+ln -s /var/lib/origin/.docker /root/
```
Comment (asalkeld) [inline pkg/arm/v10/data/master-startup.sh]: any reason this is not in the startup container?

File: pkg/arm/v10/data/node-startup.sh [hunk 5]
```diff
@@ -105,3 +112,5 @@ shutdown --reboot +2
 systemctl start atomic-openshift-node.service || true
 {{end}}
 
+# destroy yourself
+rm -rf /var/lib/waagent/custom-script/download/*/script.sh
```
Comment (asalkeld) [inline pkg/arm/v10/data/node-startup.sh]: is this to stop exposing the secret?

File: pkg/arm/v10/data/master-startup.sh (hunk 4 above)
Comment (asalkeld) [reply pkg/arm/v10/data/master-startup.sh]: ok, i'll stop being stupid. we want to pull the startup container from acr..

File: docs/release.md [hunk 6]
```diff
@@ -71,6 +71,14 @@ Branching model
    with the appropriate version on quay.io. (Note: only members of the
    `openshift-on-azure` quay.io team have access.)
 
+   Images for production are released to ACR. This can be done by people,
+   who have ACR root secrets. These images can be built by checking out 
+   `release-vX` branch and running command `make azure-push`. For this command
+   to publish images to quay.io and ACR you will need to have private secrets 
+   in your `private-secrets` directory. This can be done by executing
+   `make private-secrets`. After `private-secrets` is present, `make azure-push`
```
Comment (asalkeld) [inline docs/release.md]: make azure-push-acr ?

File: pkg/arm/v10/data/node-startup.sh (hunk 5 above)
Comment (mjudeikis) [reply pkg/arm/v10/data/node-startup.sh]: to make us sleep better :) yes

**Comment by mjudeikis** [comment]
@jim-minter, @ehashman ptal

File: pkg/arm/v10/derive.go [hunk 7]
```diff
@@ -13,3 +14,27 @@ var derived = &derivedType{}
...
+
+type DockerConfigEntry struct {
+	Auth string `json:"auth"`
+}
+
+func (derivedType) CombinedImagePullSecret(cfg *api.Config) ([]byte, error) {
+	var config DockerConfigJson
+	err := json.Unmarshal(cfg.Images.ImagePullSecret, &config)
+	if err != nil {
+		return []byte{}, err
```
Comment (jim-minter) [inline pkg/arm/v10/derive.go]: nil is traditional instead of []byte{}

File: pkg/arm/v10/derive_test.go [hunk 8]
```diff
@@ -14,3 +16,22 @@ func TestDerivedMasterLBCNamePrefix(t *testing.T) {
...
+		},
+	}
+	expected := []byte("{\"auths\":{\"osarpint.azurecr.io\":{\"auth\":\"dGVzdDp0ZXN0Cg==\"},\"registry.redhat.io\":{\"auth\":\"dGVzdDp0ZXN0Cg==\"}}}")
+
+	b, err := derived.CombinedImagePullSecret(&cfg)
+	if err != nil {
+		t.Errorf("derived.CombinedImagePullSecret() error %v", err)
+	}
+	if !reflect.DeepEqual(expected, b) {
+		t.Errorf("derived.CombinedImagePullSecret() = lenght \"%v\", want \"%v\"", len(b), len(expected))
```
Comment (jim-minter) [inline pkg/arm/v10/derive_test.go]: why not just pass string(b) and string(expected) to t.Errorf?

File: .gitignore [hunk 9]
```diff
@@ -12,6 +12,7 @@ coverage.out
 cscope.files
 cscope.out
 /secrets
+/private-secrets
```
Comment (jim-minter) [inline .gitignore]: revert

File: docs/alias.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        --from-file=./secrets/vpn-rootCA.der \
+        --from-file=./secrets/vpn-clientKey.pem \
+        --from-file=./secrets/vpn-clientCert.pem \
+        --from-file=./secrets/vpn-westeurope.ovpn \
+        --from-file=./secrets/vpn-eastus.ovpn \
+        --from-file=./secrets/vpn-australiasoutheast.ovpn \
+        --dry-run -o yaml | oc apply -n azure -f - "
+```
+
+## Acitivity logs for the cluster
```
Comment (jim-minter) [inline docs/alias.md]: remove above line 41

**Comment by jim-minter** [comment]
couple of nits here - am going to push to your branch to fix them

**Pull request merged**